runtime/onert/backend/train/ops/ConvolutionLayer.cc

```diff
@@ -24,31 +56,168 @@ namespace train
+                                  getShape(_act_back_prop_output.get()),
+                                  getBuffer<float>(_act_back_prop_output.get()));
+      backprop_act = _act_back_prop_output.get();
+      break;
+    default:
+      throw std::runtime_error("train ConvolutionLayer: Unsupported activation type yet");
+  }
+
+  // Initialize conv params for training kernels
+  nnfw::cker::ConvParams conv_train_params;
+  conv_train_params.padding_type = getPaddingType(_paddingType);
+  conv_train_params.padding_values.width = _paddingLeft;
+  conv_train_params.padding_values.height = _paddingTop;
+  conv_train_params.stride_width = _strideWidth;
+  conv_train_params.stride_height = _strideHeight;
+  conv_train_params.dilation_width_factor = _dilationWidthFactor;
+  conv_train_params.dilation_height_factor = _dilationHeightFactor;
+
+  // Transpose weights from OHWI to HWIO
+  auto transposed_weights = _transposed_weights.get();
+  assert(transposed_weights->getShape().rank() == 4);
+  nnfw::cker::TransposeParams transpose_param;
+  transpose_param.perm_count = transposed_weights->getShape().rank();
+  transpose_param.perm[0] = 1;
+  transpose_param.perm[1] = 2;
+  transpose_param.perm[2] = 3;
+  transpose_param.perm[3] = 0;
+  nnfw::cker::Transpose(transpose_param, getShape(_kernel), getBuffer<float>(_kernel),
+                        getShape(transposed_weights), getBuffer<float>(transposed_weights));
+
+  // Calculate gradient for input
+  nnfw::cker::train::ConvInputGrad(
+    conv_train_params, getShape(backprop_act), getBuffer<float>(backprop_act),
+    getShape(transposed_weights), getBuffer<float>(transposed_weights), _paddingBottom,
+    _paddingRight, getShape(_back_prop_input), getBuffer<float>(_back_prop_input));
+
+  // Calculate gradient for weights
+  auto transposed_grad_weights = _transposed_grad_weights.get();
+  assert(_grad_weights->getShape().rank() == 4);
+  assert(transposed_grad_weights->getShape().rank() == 4);
+  nnfw::cker::train::ConvFilterGrad(
+    conv_train_params, getShape(backprop_act), getBuffer<float>(backprop_act), getShape(_input),
+    getBuffer<float>(_input), _paddingBottom, _paddingRight, getShape(transposed_grad_weights),
+    getBuffer<float>(transposed_grad_weights));
+
+  // Transpose weights'gradient from HWIO to OHWI
+  nnfw::cker::TransposeParams transpose_grad_param;
+  transpose_grad_param.perm_count = transposed_grad_weights->getShape().rank();
+  transpose_grad_param.perm[0] = 3;
+  transpose_grad_param.perm[1] = 0;
+  transpose_grad_param.perm[2] = 1;
+  transpose_grad_param.perm[3] = 2;
+  nnfw::cker::Transpose(transpose_grad_param, getShape(transposed_grad_weights),
+                        getBuffer<float>(transposed_grad_weights), getShape(_grad_weights),
+                        getBuffer<float>(_grad_weights));
+
+  // Calculate gradient for bias
+  if (_bias)
+  {
+    // TODO Use optimized kernel
```

Comment:
Using this ReduceSum kernel quite affects latency. we need to apply another optimized ReduceSum kernel.